### PR TITLE
Oldify dynamically linked modules only once

### PR DESCRIPTION
**Pull request by chambart**
The list of dynamically linked modules is scanned and oldifyied completely at each minor collection. This patch avoids traversing each one more than once.

### CODE REVIEW

**Comment by damiendoligez** [comment]
This looks good. What's the best/easiest way to test it?

**Comment by chambart** [comment]
For testing it, I suppose that frama-c or jane-street code based on ocaml-plugin should torture that a bit.
For performances, the best thing to demonstrate the interest is probably the native toplevel.

**Comment by chambart** [comment]
@bobot Thanks for the question, it made me think a bit.
Originaly, my main reason for being convinced was 'it is doing the same thing as non dynlinked roots'.

Now I start to suspect that this is not even necessary to do those `Oldify` if the fields of the globals are initialized with `caml_modify` or `caml_initialize`:
If the value put inside the global is young, it is registered for major to minor pointers and will be oldified by `caml_empty_minor_heap`. Otherwise, well, it is already old.
Basically, I think that we can consider that global roots, behave like blocks in the major heap (that is, not young). @planar Am I right that we could completely remove both loops that oldify those globals ? 
Note that this part does not darken the values, it only move to the major heap.

I have absolutely no idea of the potential interaction with the aging.

**Comment by xavierleroy** [comment]
> if the fields of the globals are initialized with caml_modify or caml_initialize

But they are not, and for two good reasons:
1- caml_modify and caml_initialize assume that the destination pointer is in the minor or major heaps, and will not work if it is outside the heap (like globals are in ocamlopt);
2- calling caml_modify/initialize for every global would increase significantly the code size and compilation times of the initialization functions of compilation units.

**Comment by damiendoligez** [comment]
For aging, I had to make a slight change to the scanning of global roots, and a similar change should work for dll roots after this patch.

**Comment by chambart** [comment]
@xavierleroy effectively, I didn't consider the second point.
Yet for the first point I think this works correctly for the globals since they are scanned at the beginning of the major cycle (contrarily to other out of heap blocks). I may be wrong, but I can't see where this would be any different from a block in the major heap.
This does not matter in this case, but I am concerned about that for this other pull request https://github.com/ocaml/ocaml/pull/178 where the invariant that a global block won't be modified twice doesn't hold.

**Comment by damiendoligez** [comment]
Tested on frama-c, no bug found.

**Comment by chambart** [comment]
Thanks for the tests. I couldn't find bugs with my testsuite either.
I think that the invariant we may want to ensure is that the globals are not modified once initialized. Would you want a patch ensuring that through mprotect (in debug-runtime mode) (of course also disabling it during a compaction) ?

**Comment by damiendoligez** [comment]
It would be useful, but don't forget I also use debug-runtime on Mac OS and Windows...

**Comment by chambart** [comment]
@bobot I think it does not need any modification anymore.
The sugested improved assertions proved to be a bit too tricky to implement (marshal can modify immutable values), so it's not going to be done.

**Comment by alainfrisch** [comment]
Since this an optimization, can you provide some estimate of the expected gains (or even just a micro-benchmark)?

**Comment by chambart** [comment]
edit: original message is an error, I mixed with pr https://github.com/ocaml/ocaml/pull/156

@alainfrisch On a quite agressively inlined frama-c this was ~0.3s on a ~1s analysis
On my stupid micro benchmarks for testing that, this could be made arbitrarily beneficial.
In practice I do not know any other software doing as much dynlinking as frama-c, but I imagine that on ocamlnat, it would be hugely beneficial. (I didn't took the time to fix ocamlnat it to test it)

**Comment by yakobowski** [comment]
@chambart  Did you notice any improvement when starting Frama-C without performing any analysis? Startup times are increasing (which is a problem when running our unit tests), and I'm wondering if this PR might improve things.

**Comment by alainfrisch** [comment]
> On a quite agressively inlined frama-c this was ~0.3s on a ~1s analysis

That sounds great.  (LexiFi's application also dynlinks a large part of its code at startup time, so we might see good improvements as well.)

Just curious: how does the "agressively inlined" factor interact with the current proposal?

**Comment by alainfrisch** [comment]
But this proposal is about the list of global values, not frame tables, and there is exactly one global per dynlinked module, no?

(Well I see in #177 that you also propose to support multiple globals per unit.)

**Comment by chambart** [comment]
oups, sorry @alainfrisch and @yakobowski, I'm mixing this pr with https://github.com/ocaml/ocaml/pull/156.
Sorry I don't have numbers ready for this one.

**Comment by alainfrisch** [comment]
Ok thanks!  My intuition is that the gain for this one would be tiny, but the change is tiny as well.  I'm in favour of merging.

That said, in general, I strongly advise to provide numbers when proposing an optimization (preferably with realistic benchmarks or even better, real programs; but micro-benchmarks at least otherwise); and also intuitions about when the optimization would be beneficial.  This gives some more incentive to look at the proposal.

**Comment by chambart** [comment]
I just retested (for benchmarking) frama-c with this pr and it segfaulted. I don't know what causes this right now.

**Comment by gasche** [comment]
Any updates on the segfault? It looks like #178 would be mergeable, but I'm waiting on this one first.

**Comment by radekm** [comment]
> Any updates on the segfault? It looks like #178 would be mergeable, but I'm waiting on this one first.

Wasn't the segfault related to [PR#6919: corrupted final_table](http://caml.inria.fr/mantis/view.php?id=6919)?

**Comment by mshinwell** [comment]
> Wasn't the segfault related to PR#6919: corrupted final_table?

No, says pchambart

**Comment by damiendoligez** [comment]
@chambart two questions:
1. any news of the segfault?
2. is this a prerequisite for flambda?

**Comment by chambart** [comment]
Not a prerequisite of flambda, and I know how to fix that, but it is a bit more complex than this patch and I don't really have time for that right now.

**Comment by damiendoligez** [comment]
I'm postponing this PR to 4.04. Speak up if you disagree.

**Comment by chambart** [comment]
@bobot, the code here is wrong, and I would still need some time on it. I'm posponing.

**Comment by xavierleroy** [comment]
Any news on this one?

**Comment by mshinwell** [comment]
I'm going to close this; there hasn't been any activity for a year and (forgive me) there are many more important things for @chambart to be getting on with :)
